circuit_knitting/cutting/cutting_decomposition.py

```diff
@@ -260,3 +267,105 @@ def decompose_observables(
+
+    Args:
+        circuit: The circuit to cut
+        optimization: Settings for controlling optimizer behavior. Currently,
+            only a best-first optimizer is supported.
+        constraints: QPU constraints used to generate the cut location search space.
+
+    Returns:
+        A circuit containing :class:`.BaseQPDGate` instances. The subcircuits
+        resulting from cutting these gates will be runnable on the devices
+        specified in ``constraints``.
+
+        A metadata dictionary:
+            cuts: A list of length-2 tuples describing each cut in the output circuit.
+                The tuples are formatted as ``(cut_type: str, cut_id: int)``. The
+                cut ID is the index of the cut gate or wire in the output circuit's
+                ``data`` field.
+            sampling_overhead: The sampling overhead incurred from cutting the specified
+                gates and wires.
+    """
+    circuit_cco = qc_to_cco_circuit(circuit)
+    interface = SimpleGateList(circuit_cco)
+
+    opt_settings = OptimizationSettings.from_dict(optimization)
+
+    # Hard-code the optimization type to best-first
+    opt_settings.set_engine_selection("CutOptimization", "BestFirst")
+
+    constraint_settings = DeviceConstraints.from_dict(constraints)
+
+    # Hard-code the optimizer to an LO-only optimizer
+    optimizer = LOCutsOptimizer(interface, opt_settings, constraint_settings)
+
+    # Find cut locations
+    opt_out = optimizer.optimize()
+
+    wire_cut_actions = []
+    gate_ids = []
+
+    opt_out = cast(DisjointSubcircuitsState, opt_out)
+    opt_out.actions = cast(list, opt_out.actions)
+    for action in opt_out.actions:
+        if action[0].get_name() == "CutTwoQubitGate":
+            gate_ids.append(action[1][0])
+        else:
+            # We only support four types of optimizer cuts
+            assert action[0].get_name() in (
+                "CutLeftWire",
+                "CutRightWire",
+                "CutBothWires",
+            )
+            wire_cut_actions.append(action)
+
+    # First, replace all gates to cut with BaseQPDGate instances.
+    # This assumes each gate to cut is replaced 1-to-1 with a QPD gate.
+    # This may not hold in the future as we stop treating gate cuts individually.
+    circ_out = cut_gates(circuit, gate_ids)[0]
+
+    # Insert all the wire cuts
+    counter = 0
+    for action in sorted(wire_cut_actions, key=lambda a: a[1][0]):
+        inst_id = action[1][0]
+        # action[2][0][0] will be either 1 (control) or 2 (target)
+        qubit_id = action[2][0][0] - 1
+        circ_out.data.insert(
+            inst_id + counter,
+            CircuitInstruction(
+                CutWire(), [circuit.data[inst_id + counter].qubits[qubit_id]], []
+            ),
+        )
+        counter += 1
+        if action[0].get_name() == "CutBothWires":
+            # qubit_id is 0 or 1. Flip it.
```

Comment:
This may be a CKT/Qiskit thing I am unfamiliar with but why do we do this flipping? Does it have to do with qiskit using little endian notation?

Reply:
No, I'll clarify this in the comment. Basically I'm just saying that if this is a "CutBothWires", we have a second wire in our "Action" to cut and the gate-relative qubit for the second wire will be the opposite of whatever the first qubit was.

We know qubit_id is 0 or 1 here because we only support 2-qubit gates, so we just flip it and cut that one too. We don't actually need to read it from the structure, althought I suppose we could for clarity

Reply:
OK we now read qubit_id2 from the data structure as well instead of this funky logic. It's more clear what is going on.

Reply:
Yes, thanks.